### PR TITLE
fix(release): publish validated commits and schedule safety checks

### PR DESCRIPTION
**Pull request by 0xpaulx**
No linked issue.

#### Describe the changes you have made in this PR -

- gate rolling releases on a successful `main` CI workflow run and build the validated source SHA
- preserve the existing release path exclusions through a tested classifier
- run the release safety regression tests remotely every hour at minute 17 UTC

### Demo/Screenshot for feature changes and bug fixes -

```text
uv run pytest tests/github_ci/test_release_safety.py
3 passed in 0.45s

make pre-push
Validation: 9/9 passed in 13.2s.
```

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

The release workflow now consumes the successful CI workflow's `head_sha`, checks out that exact commit in every build and publish job, and skips rolling releases for the same paths previously excluded by the push trigger. A small Python classifier keeps that path decision testable. The separate least-privilege scheduled workflow installs the locked development environment and runs the focused release safety tests hourly. It also supports manual dispatch for deployment verification.

A direct scheduled release was rejected because it could publish an unvalidated commit. Reusing the full CI matrix hourly was also rejected because the focused release contracts provide the needed signal at much lower runner cost.

---

## Checklist before requesting a review
- [x] I have added a proper PR title; no issue was provided to link
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -55,20 +55,38 @@ jobs:
...
           fi
 
+          if ! git cat-file -e "${source_sha}^{commit}"; then
+            echo "Release source commit is unavailable: ${source_sha}" >&2
+            exit 1
+          fi
+
+          if [ "$EVENT_NAME" = "workflow_run" ]; then
+            changed_paths="$(git diff --name-only "${source_sha}^1" "$source_sha")"
+            if [ "$(printf '%s\n' "$changed_paths" | python3 .github/ci/release_paths.py)" != "true" ]; then
```
[the remark this answers is not in the document]
Comment (0xpaulx) [reply .github/workflows/release.yml:93]: Fixed in 5a4b388. The classifier now compares the last published main-build tag with the validated source SHA, so it covers every unreleased commit rather than only the immediate parent. It also skips stale workflow completions and handles the first publication by classifying the full source tree.

File: .github/workflows/release-safety-scheduled.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  release-safety:
+    name: release-safety
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+
+    steps:
+      - uses: actions/checkout@v5
+
+      - name: Install uv
+        uses: astral-sh/setup-uv@v10.0.1
```
[the remark this answers is not in the document]
Comment (0xpaulx) [reply .github/workflows/release-safety-scheduled.yml]: Fixed in 5a4b388. checkout v5, setup-uv v10.0.1, and setup-python v6 are now pinned to their resolved full commit SHAs, with version comments retained for maintenance.

**Comment by 0xpaulx** [comment]
@greptile review

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -55,20 +55,49 @@ jobs:
...
           if [ "$EVENT_NAME" = "workflow_dispatch" ] && [ "$DISPATCH_CHANNEL" = "main" ]; then
             channel="main"
           fi
 
+          if ! git cat-file -e "${source_sha}^{commit}"; then
+            echo "Release source commit is unavailable: ${source_sha}" >&2
+            exit 1
+          fi
+
+          if [ "$EVENT_NAME" = "workflow_run" ]; then
```
[the remark this answers is not in the document]
Comment (0xpaulx) [reply .github/workflows/release.yml:80]: Fixed in 3936da8. The release upload/create/edit step now completes before main-build is moved. The publishing step also uses MAIN_BUILD_TAG_TOKEN when available, so initial release creation can target the validated SHA. If publication fails, the baseline remains unchanged and the unreleased range is retried.

**Comment by 0xpaulx** [comment]
@greptile review
